<font color='blue'> Привет. Поздравляю с первым самостоятельным проектом. Ниже мои комментарии выделены синим. </font>

## Исследование надёжности заёмщиков


### Шаг 1. Откроем файл с данными и изучим общую информацию. 

In [153]:
import pandas as pd
df=pd.read_csv('/datasets/data.csv')
df.info()
df.head(15)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

### Вывод

- В таблице присутствуют целочисленные, строковые и вещественные значения
- Столбец "days_employed" и "total_income" содержит пропуски 
- Столбец "education" содержит дубликаты с прописным и заглавным написанием 
- Столбец "purpose" содержит смысловые дубликаты


<font color='blue'> Хорший анализ сырых данных. </font>

### Шаг 2. Предобработка данных

### Обработка пропусков

In [154]:
#Вглянем на строки таблицы с пропусками

print(df[df['total_income'].isnull()].head())

    children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29   

Пропускам в столбце "days_employed" соотвествуют пропуски в "total_income"


In [155]:
#Найдем средние значения для столбцов "days_employed" и "total_income"

mean_days_employed=df['days_employed'].mean()
print("Средний стаж работы ",mean_days_employed)
mean_total_income=df['total_income'].mean()
print("Средний доход ",mean_total_income)

Средний стаж работы  63046.49766147338
Средний доход  167422.30220817294


In [156]:
# и заменим ими пропуски

df['days_employed']=df['days_employed'].fillna(mean_days_employed)
df['total_income']=df['total_income'].fillna(mean_total_income)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<font color='blue'> Хоршо, пропуски заполнены верно. </font>

### Вывод

- Пропускам в столбце "total_income" соответствуют пропуски в столбце "days_employed".
  Это может быть связано с ошибкой ввода оператора или техническим сбоем программы
- Пропуски были заменены средними значениями по столбцу

### Замена типа данных

In [157]:
df['days_employed']=df['days_employed'].astype('int')
df['total_income']=df['total_income'].astype('int')
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years education  education_id  \
0         1          -8437         42    высшее             0   
1         1          -4024         36   среднее             1   
2         0          -5623         33   Среднее             1   
3         3          -4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

<font color='blue'> Типы данных заменены правильно. </font>

### Вывод

- Тип  float64 заменили на тип int64, так как дробная часть нас не интересует

### Обработка дубликатов

Узнаем какие уникальные значения имеет столбец 'education'

In [158]:
print(df['education'].unique(),'\n')

['высшее' 'среднее' 'Среднее' 'СРЕДНЕЕ' 'ВЫСШЕЕ' 'неоконченное высшее'
 'начальное' 'Высшее' 'НЕОКОНЧЕННОЕ ВЫСШЕЕ' 'Неоконченное высшее'
 'НАЧАЛЬНОЕ' 'Начальное' 'Ученая степень' 'УЧЕНАЯ СТЕПЕНЬ'
 'ученая степень'] 



Узнаем какие уникальные значения имеет столбец 'family_status'

In [159]:
print(df['family_status'].unique(),'\n')

['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'Не женат / не замужем'] 



Узнаем какие уникальные значения имеет столбец 'income_type'

In [160]:
print(df['income_type'].unique(),'\n')

['сотрудник' 'пенсионер' 'компаньон' 'госслужащий' 'безработный'
 'предприниматель' 'студент' 'в декрете'] 



Устраним повторы в столбце 'education' приведением всех значений к нижнему регистру

In [161]:
df['education']=df['education'].str.lower()

Вглянем на значения в столбце children     

In [162]:
print(df['children'].value_counts())

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64


Так как дефектных данных со значениями "-1" не так уж много просто удалим эти строки 

In [163]:
df = df[df['children']>-1]
print(df.head())

   children  days_employed  dob_years education  education_id  \
0         1          -8437         42    высшее             0   
1         1          -4024         36   среднее             1   
2         0          -5623         33   среднее             1   
3         3          -4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  


Найдем полные дубликаты в строках таблицы

In [164]:
print(df.duplicated().count())

21478


Удалим дубликаты и проверим количество строк в таблице

In [165]:
df.drop_duplicates()
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21478 entries, 0 to 21524
Data columns (total 12 columns):
children            21478 non-null int64
days_employed       21478 non-null int64
dob_years           21478 non-null int64
education           21478 non-null object
education_id        21478 non-null int64
family_status       21478 non-null object
family_status_id    21478 non-null int64
gender              21478 non-null object
income_type         21478 non-null object
debt                21478 non-null int64
total_income        21478 non-null int64
purpose             21478 non-null object
dtypes: int64(7), object(5)
memory usage: 2.1+ MB
None


Теперь строк осталось только 21478

### Вывод

- Дубликаты в столбце 'education' были найдены пониженим регистра
- Дефектные данные из столбца children  удалены
- Полностью задублированные строки также удалены

<font color='blue'> Здесь обработка дубликатов немного не доделана. Надо посчитать количество дубликатов, используя метод duplicated(), потом надо удалить их и снова посчитать (должно стать 0). </font>

<font color='red'> Ок. </font>

### Лемматизация

Узнаем какие уникальные значения имеет столбец 'purpose'

In [166]:
print(df['purpose'].unique())

['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']


Лемматизируем все полученнные значения и оставим только уникальные существительные

In [167]:

from pymystem3 import Mystem
m = Mystem()
from collections import Counter

#Добавим в список lemmas лемматизированные слова из столбца 'purpose'
lemmas=[]            
for row in df['purpose'].unique(): 
    for word in m.lemmatize(row):       
        lemmas.append(word)
        
#Добавим в список lemmas_list только уникальные слова (кроме предлогов и тд) из списка lemmas      
lemmas_list=[]
for item in lemmas: 
    if (item not in lemmas_list) and (len(item)>4): 
        lemmas_list.append(item)    
        
lemmas_list

['покупка',
 'жилье',
 'приобретение',
 'автомобиль',
 'дополнительный',
 'образование',
 'сыграть',
 'свадьба',
 'операция',
 'проведение',
 'семья',
 'недвижимость',
 'коммерческий',
 'жилой',
 'строительство',
 'собственный',
 'подержать',
 'заниматься',
 'сделка',
 'подержанный',
 'получение',
 'высокий',
 'профильный',
 'сдача',
 'ремонт']

In [168]:
#Вручную определим список лемм
lemma=['жилье','автомобиль','образование','свадьба','недвижимость','ремонт'] 

<font color='blue'> Операция - этот странная лемма для категорий. С учетом того, что обычно пишут - опреация с недвижимостью, т.е это недвижимость.... </font>

Категории из набора lemma встречаются в каждой строчке и ёмко отражают цель кредита

In [169]:
#С помощью функции goal зададим каждой цели свою категорию
def goal(purpose):
    for word in lemma:
        if word in m.lemmatize(purpose):
            return word
        
#Создадим дополнительный столбец "goal" для категоризации цели кредита        
df['goal'] = df['purpose'].apply(goal) 

print(df.head(10))
df.info()

   children  days_employed  dob_years education  education_id  \
0         1          -8437         42    высшее             0   
1         1          -4024         36   среднее             1   
2         0          -5623         33   среднее             1   
3         3          -4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0           -926         27    высшее             0   
6         0          -2879         43    высшее             0   
7         0           -152         50   среднее             1   
8         2          -6929         35    высшее             0   
9         0          -2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M

<font color='blue'> Вот здесь операция - это лишняя категория, она перетянула часть от жилья и недвижимости. </font>

In [170]:
df['goal'].value_counts()

недвижимость    6351
жилье           4465
автомобиль      4304
образование     4012
свадьба         2346
Name: goal, dtype: int64

<font color='red'> Ок. </font>

### Вывод

Из большого числа целей кредита были выделены характерные леммы.
На их основе был создан дополнтельный столбец с шестью категориям целей кредита.

<font color='blue'> Сама лемматизаци выполнена верно, но категории выделены не совсем правильно. </font>

### Категоризация данных

In [171]:
#Узнаем максимальный уровень доходов
print('Максимальный доход ',df['total_income'].max())
#и минимальный
print('Минимальный доход ',df['total_income'].min())

Максимальный доход  2265604
Минимальный доход  20667


<font color='blue'> Для категорий по доходу нужно использовать квантили, а не просто произвольные уровни. </font>

Разобьем доходы на 3 группы с помощью квантилей

In [172]:
print(df['total_income'].quantile([0.33,0.66]))

0.33    122313.97
0.66    169021.10
Name: total_income, dtype: float64


<font color='red'> Ок. </font>



Добавим специальный столбец с уровнем доходов для легкой группировки в дальнейшем

Примем доход до 122000 руб за низкий
Доход от 122313 до 169000 за средний
и доход свыше 169021 за высокий

In [173]:
def income_level(inc):
    if inc<=122000:
        return 'низкий'
    if 12200<inc<=169000:
        return 'средний'
    if inc>169000:
        return 'высокий'
    
df['income_level'] = df['total_income'].apply(income_level) 
print(df.head(10))

   children  days_employed  dob_years education  education_id  \
0         1          -8437         42    высшее             0   
1         1          -4024         36   среднее             1   
2         0          -5623         33   среднее             1   
3         3          -4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0           -926         27    высшее             0   
6         0          -2879         43    высшее             0   
7         0           -152         50   среднее             1   
8         2          -6929         35    высшее             0   
9         0          -2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M

<font color='blue'> Категоризировать по количеству детей и по семейному положению не надо - это сокращает информативность данных. </font>

### Вывод

Были выделены три группы доходов. Далее по ним можно будет группировать данные.

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [174]:
#Сгруппируем таблицу по столбцу 'children', при этом подсчитаем сумму должников 
# и общее количество участников для кадой группы методами sum и count 
children_grouped = df.groupby('children').agg({'debt': ['count', 'sum']})

#Найдем процентное соотношение должников для каждой группы. Результат запишем в столбец "procent"
children_grouped['procent']=children_grouped['debt']['sum']/children_grouped['debt']['count']*100
print(children_grouped.sort_values(by='procent'))

           debt          procent
          count   sum           
children                        
5             9     0   0.000000
0         14149  1063   7.512898
3           330    27   8.181818
1          4818   444   9.215442
2          2055   194   9.440389
4            41     4   9.756098
20           76     8  10.526316


<font color='blue'> В целом правильно, но надо убрать категоризацию, хотелось бы видить проценты для каждого количества детей. </font>

<font color='red'> Ок. </font>

### Вывод

Наличие детей увеличивает риск просроченных платежей 

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [175]:
#Сгруппируем таблицу по столбцу 'family_status', при этом подсчитаем сумму должников 
# и общее количество участников для кадой группы методами sum и count 
family_status_grouped = df.groupby('family_status').agg({'debt': ['count', 'sum']})

#Найдем процентное соотношение должников для каждой группы. Результат запишем в столбец "procent"
family_status_grouped['procent']=family_status_grouped['debt']['sum']/family_status_grouped['debt']['count']*100
print(family_status_grouped.sort_values(by='procent'))

                        debt        procent
                       count  sum          
family_status                              
вдовец / вдова           956   63  6.589958
в разводе               1191   85  7.136860
женат / замужем        12351  930  7.529755
гражданский брак        4172  388  9.300096
Не женат / не замужем   2808  274  9.757835


<font color='blue'> Здесь аналогично. </font>

### Вывод

Наличие партнера увеличивает шансы возврата кредита 

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [176]:
#Сгруппируем таблицу по столбцу 'income_level', при этом подсчитаем сумму должников 
# и общее количество участников для кадой группы методами sum и count 
income_grouped = df.groupby('income_level').agg({'debt': ['count', 'sum']})

#Найдем процентное соотношение должников для каждой группы. Результат запишем в столбец "procent"
income_grouped['procent']=income_grouped['debt']['sum']/income_grouped['debt']['count']*100
print(income_grouped.sort_values(by='procent'))

              debt        procent
             count  sum          
income_level                     
высокий       7305  556  7.611225
низкий        7043  577  8.192532
средний       7130  607  8.513324


### Вывод

Людис высоким доходом чаще возвращают кредит в срок

- Как разные цели кредита влияют на его возврат в срок?

In [177]:
#Сгруппируем таблицу по столбцу 'income_level', при этом подсчитаем сумму должников 
# и общее количество участников для кадой группы методами sum и count 
goal_grouped = df.groupby('goal').agg({'debt': ['count', 'sum']})

#Найдем процентное соотношение должников для каждой группы. Результат запишем в столбец "procent"
goal_grouped['procent']=goal_grouped['debt']['sum']/goal_grouped['debt']['count']*100
print(goal_grouped.sort_values(by='procent'))

              debt        procent
             count  sum          
goal                             
жилье         4465  308  6.898096
недвижимость  6351  474  7.463392
свадьба       2346  186  7.928389
образование   4012  370  9.222333
автомобиль    4304  402  9.340149


### Вывод

Наиболее часто не возвращают кредит вовремя на авто, затем на образование. 
Реже встречаются задолженности по кредиту на жилье и недвижимость.

### Шаг 4. Общий вывод

- Обнаружена зависимость между наличием детей и возвратом кредита в срок
- Обнаружена зависимость между семейным положением и возвратом кредита в срок
- Обнаружена зависимость между уровнем дохода и возвратом кредита в срок
- Обнаружена зависимость между целями кредита и возвратом кредита в срок

<font color='blue'> В целом проект вполнен не плохо. постарайся доработать по замечаниям. Удачи. </font>

<font color='red'> Проект отлично доработан, молодец. </font>

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.